net/dns/wsl_windows.go

```diff
@@ -197,21 +198,75 @@ func (fs wslFS) WriteFile(name string, contents []byte, perm os.FileMode) error
 	cmd := fs.cmd("tee", "--", name)
 	cmd.Stdin = bytes.NewReader(contents)
 	cmd.Stdout = nil
-	if err := cmd.Run(); err != nil {
+	if err := wslRun(cmd); err != nil {
 		return err
 	}
-	return fs.cmd("chmod", "--", fmt.Sprintf("%04o", perm), name).Run()
+	return wslRun(fs.cmd("chmod", "--", fmt.Sprintf("%04o", perm), name))
 }
 
 func (fs wslFS) cmd(args ...string) *exec.Cmd {
 	cmd := wslCommand("-u", fs.user, "-d", fs.distro, "-e")
 	cmd.Args = append(cmd.Args, args...)
-	fmt.Printf("wslFS.cmd: %v\n", cmd.Args)
 	return cmd
 }
 
 func wslCommand(args ...string) *exec.Cmd {
 	cmd := exec.Command("wsl.exe", args...)
-	cmd.SysProcAttr = &syscall.SysProcAttr{HideWindow: true}
 	return cmd
 }
+
+func wslCombinedOutput(cmd *exec.Cmd) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	cmd.Stdout = buf
+	cmd.Stderr = buf
+	err := wslRun(cmd)
+	return buf.Bytes(), err
+}
+
+func wslRun(cmd *exec.Cmd) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("wslRun(%v): %v", cmd.Args, err)
+		}
+	}()
+
+	var token syscall.Token
+	if u, err := user.Current(); err == nil && u.Name == "SYSTEM" {
+		// We need to switch user to run wsl.exe.
+		// https://github.com/microsoft/WSL/issues/4803
+		sessionID, err := wtsGetActiveConsoleSessionID()
+		if err != nil {
+			return err
+		}
+
+		var t windows.Token
+		if err := windows.WTSQueryUserToken(sessionID, &t); err != nil {
+			return err
+		}
+		defer t.Close()
+		token = syscall.Token(token)
+	}
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Token:      token,
+		HideWindow: true,
+	}
+	return cmd.Run()
+}
+
+var (
+	kernel32                         = syscall.NewLazyDLL("kernel32.dll")
+	procWTSGetActiveConsoleSessionId = kernel32.NewProc("WTSGetActiveConsoleSessionId")
+)
+
+// wtsGetActiveConsoleSessionID Returns the session ID of the console session.
+//
+// https://msdn.microsoft.com/en-us/library/aa383835(VS.85).aspx
+// DWORD WTSGetActiveConsoleSessionId();
+func wtsGetActiveConsoleSessionID() (sessionID uint32, err error) {
+	r1, _, e1 := procWTSGetActiveConsoleSessionId.Call()
+	if r1 == 0xFFFFFFFF {
```

Comment:
check e1 too? Oh, I guess API docs are pretty clear that this is enough.

Reply:
The Proc.Call method is bonkers:

_The returned error is always non-nil, constructed from the result of GetLastError. Callers must inspect the primary return value to decide whether an error occurred (according to the semantics of the specific function being called) before consulting the error._

https://pkg.go.dev/golang.org/x/sys/windows#Proc.Call